---
title: Check RNA-seq strandedness
weight: 1

categories:

tags:

authors:
  - christopher-barrington

type: docs
toc: true
draft: false

date: `r format(Sys.time(), "%Y-%m-%d")`
created: {{ now.Format "2006-01-02" }}

always_allow_html: true
output:
  html_document:
    keep_md: yes
    df_print: paged
---

<!-- initial set up chunks are invisible -->

<!-- make subdirectories to write into/any bash work -->
```{bash setup_wd, opts.label='batch_chunk', code=readLines(system.file('chunks/working_directory.sh', package='project.notes.tools'))}
```

<!-- read the standard knitr setup script -->
```{r knitr_setup, opts.label='batch_chunk', code=readLines(system.file('chunks/knitr.r', package='project.notes.tools'))}
```

<!-- load any packages for the workbook -->
```{r load_workbook_packages, opts.label='batch_chunk'}
library(yaml)
```

<!-- read the standard setup script, loading well-used libraries -->
```{r configure_r, opts.label='batch_chunk', code=readLines(system.file('chunks/environment.r', package='project.notes.tools'))}
```

<!-- DOIs for software used in the workbook or to generate presented results, a named vector of strings -->
```{r add_doi_to_bibliography, cache=TRUE, echo=FALSE}
c(RSeQC='10.1093/bioinformatics/bts356',
  STAR='10.1093/bioinformatics/bts635') -> academic_doi
c() -> software_doi
```

<!-- write a bibliography -->
```{r write_bibliography, dependson='add_doi_to_bibliography', cache=TRUE, echo=FALSE, code=readLines(system.file('chunks/write_bibliography.r', package='project.notes.tools'))}
```

<!-- set a ggplot theme to use as a base in the workbook -->
```{r define_global_ggplot_theme, opts.label='batch_chunk'}
(theme_bw() +
 theme(text=element_text(size=10, colour='black'),
       panel.grid.minor=element_blank(),
       strip.background=element_blank())) %>%
theme_set()
```

<!-- any final things to set for this workbook -->
```{r final_setup, opts.label='batch_chunk'}
opts_chunk$set(cache=FALSE)
result_exists <- file.exists('files/infer_experiment.result.txt')
```

<!-- define slugs for other reports here for ease and reuse -->
```{r define_slugs, opts.label='batch_chunk'}
slugs <- list()
```

<!-- define shortcut links -->
[CRICK]: https://www.crick.ac.uk
[STAR]: https://academic.oup.com/bioinformatics/article/29/1/15/272537
[RSEQC]: https://academic.oup.com/bioinformatics/article/28/16/2184/325191
[INFER EXPERIMENT HELP]: http://rseqc.sourceforge.net/#infer-experiment-py
<!-- end of initial set up chunks -->

The strandedness of the RNA-seq library will be checked here using a sample of reads and the [RSeQC][RSeQC] package.

<!--more-->

## Load required project information

First, we need to load the genome specification from the data repository. These values will be exported and used to align to the correct genome index.

```{r load_genome}
system.file('docs/genome.yaml', package='datarepository') %>%
  read_yaml() -> genome
```

```{r show_genome}
genome %>%
  enframe(name='', value='') %>%
  kableExtra::kable(format=if_else(interactive(), 'markdown', 'html'))
```

Second, we need to determine whether this project has a single- or paired-end data.

```{r get_fastq}
system.file('primary_data', package='datarepository') -> fastq_path

list.files(path=fastq_path, pattern='*fastq.gz', recursive=TRUE, full.names=FALSE) %>%
  as.data.frame() %>%
  set_names('file') %>%
  extract(col=file, regex='(.+)/fastq/(.+?_S.+?_L.+?)_(R.)_.*', into=c('flowcell','sample','read.pair'), remove=FALSE) %>%
  dlply(~flowcell+sample) %>%
  pluck({length(.) %>% seq() %>% sample(, size=1)}, 'file') %>%
  file.path(fastq_path, .) -> fastq

length(fastq) %>%
  as.character() %>%
  switch(`1`='single_end',
         `2`='paired_end',
         knit_exit(append='cannot determine pairedness', fully=TRUE)) -> pairedness
```

Now the variables can be exported ready to be used in the analysis.

```{r export_variables}
list(SPECIES=genome$species,
     GENOME_SOURCE=genome$source,
     GENOME_VERSION=genome$version,
     GENOME_RELEASE=genome$release,
     GENOME_PATH=file.path('/camp/svc/reference/Genomics/babs', genome$species, genome$source, genome$version, str_c('release-', genome$release)),
     R1_FASTQ_FILE=fastq[1]) %>%
  when(pairedness=='single_end'~.,
       pairedness=='paired_end'~append(., list(R2_FASTQ_FILE=fastq[2]))) %>%
  do.call(what=Sys.setenv)
```

## Determine strandedness

With the above arguments, the strandedness can be determined using [RSeQC][RSEQC], and [STAR][STAR] for the alignment.

### Subsample FastQ files

The first 1 million records of the FastQ file(s) are extracted and saved to new files. This could be done any way, but because the records are in a random order taking the first records should be fine.

```{bash subsample_fastq_single, eval=pairedness=='single_end' & !result_exists, echo=pairedness=='single_end'}
gunzip --to-stdout ${R1_FASTQ_FILE} | head --lines 4000000 > scratch/subsample_R1.fastq
```

```{bash subsample_fastq_paired, eval=pairedness=='paired_end' & !result_exists, echo=pairedness=='paired_end'}
gunzip --to-stdout ${R1_FASTQ_FILE} | head --lines 4000000 > scratch/subsample_R1.fastq
gunzip --to-stdout ${R2_FASTQ_FILE} | head --lines 4000000 > scratch/subsample_R2.fastq
```

### Align subsampled records

[STAR][STAR] is used to align the subsampled data to the genome. The version of [STAR][STAR] used here is hard-coded and is `2.5.2a`.

```{bash run_star_single, eval=pairedness=='single_end' & !result_exists, echo=pairedness=='single_end'}
module load STAR/2.5.2a-foss-2016b

STAR \
  --runThreadN 6 \
  --genomeDir ${GENOME_PATH}/genome_idx/star/100bp/ \
  --readFilesIn scratch/subsample_R1.fastq \
  --outFileNamePrefix scratch/subsample. \
  --outSAMunmapped Within \
  --outFilterType BySJout \
  --outSAMattributes NH HI AS NM MD \
  --outFilterMultimapNmax 20 \
  --outFilterMismatchNmax 999 \
  --outFilterMismatchNoverReadLmax 0.04 \
  --alignIntronMin 20 \
  --alignIntronMax 1000000 \
  --alignMatesGapMax 1000000 \
  --alignSJoverhangMin 8 \
  --alignSJDBoverhangMin 1 \
  --sjdbScore 1 \
  --outSAMtype BAM SortedByCoordinate 
```

```{bash run_star_paired, eval=pairedness=='paired_end' & !result_exists, echo=pairedness=='paired_end'}
module load STAR/2.5.2a-foss-2016b

STAR \
  --runThreadN 6 \
  --genomeDir ${GENOME_PATH}/genome_idx/star/100bp/ \
  --readFilesIn scratch/subsample_R1.fastq scratch/subsample_R2.fastq \
  --outFileNamePrefix scratch/subsample. \
  --outSAMunmapped Within \
  --outFilterType BySJout \
  --outSAMattributes NH HI AS NM MD \
  --outFilterMultimapNmax 20 \
  --outFilterMismatchNmax 999 \
  --outFilterMismatchNoverReadLmax 0.04 \
  --alignIntronMin 20 \
  --alignIntronMax 1000000 \
  --alignMatesGapMax 1000000 \
  --alignSJoverhangMin 8 \
  --alignSJDBoverhangMin 1 \
  --sjdbScore 1 \
  --outSAMtype BAM SortedByCoordinate 
```

### Analyse the alignments

[RSeQC][RSEQC] is now used to analyse the aligned reads to determine the orientation of alignments with respect to direction of transcription in the gene models. The version of [RSeQC][RSEQC] used here is hard-coded and is `2.6.4`. We use the `infer_experiment.py` program from [RSeQC][RSEQC] which is [described in detail online][INFER EXPERIMENT HELP].


```{bash run_infer_experiment, eval=!result_exists}
module load RSeQC/2.6.4-foss-2016b-Python-2.7.12-R-3.3.1

REFGENE=${GENOME_PATH}/gtf/${SPECIES^}.${GENOME_VERSION}.${GENOME_RELEASE}.bed

infer_experiment.py \
  --input-file scratch/subsample.Aligned.sortedByCoord.out.bam \
  --refgene ${REFGENE} > files/infer_experiment.result.txt
```

The output of `infer_experiment.py` is shown below:

```{r show_results, echo=FALSE}
readLines('files/infer_experiment.result.txt') %>%
  str_subset('^$', negate=TRUE) %>%
  str_c('\n') %>%
  cat()
```

For each pair of symbols (`++` for example), the symbols show the orientation of the alignment and the direction of transcription. There are two pieces of information to consider from these results:

1. The proportion of reads in either class
2. The orientation of the read to the transcript

```{r reformat_results, echo=FALSE}
readLines('files/infer_experiment.result.txt') %>%
  str_subset(':') %>%
  str_remove_all('"') %>%
  str_remove('Fraction of reads (explained by )?') %>%
  as.data.frame() %>%
  set_names('result') %>%
  separate(col=result, into=c('alignment.class', 'proportion'), sep=': ', convert=TRUE) %>%
  mutate(percent=percent(proportion, accuracy=0.1)) %>%
  mutate(interpretation=case_when(alignment.class=='++,--' ~ 'forward:read aligns in direction of transcription',
                                  alignment.class=='+-,-+' ~ 'reverse:read aligns opposite to direction of transcription',
                                  alignment.class=='1++,1--,2+-,2-+' ~ 'forward:read 1 aligns in direction of transcription',
                                  alignment.class=='1+-,1-+,2++,2--' ~ 'reverse:read 2 aligns in direction of transcription',
                                  TRUE~'ambiguous:reads aligned to regions that cannot be used')) %>%
  separate(col=interpretation, into=c('interpretation', 'description'), sep=':', remove=FALSE) %>%
  relocate(percent, .after=last_col()) %T>%
  kable(x=select(., -proportion), format=if_else(interactive(), 'markdown', 'html')) -> results
```

```{r get_conclusion, echo=FALSE}
results %>%
  arrange(proportion) %>%
  tail(n=1) %>%
  select(proportion, interpretation) %>%
  mutate(protocol=case_when(proportion>=0.7 ~ 'stranded',
                            between(proportion, 0.4, 0.6) ~ 'unstranded',
                            TRUE~'error')) %>%
  as.list() %>%
  when(.$protocol=='error'~'The strandedness could not be determined!',
       .$protocol=='stranded'~sprintf(fmt='The protocol is stranded and %s.', .$interpretation),
       .$protocol=='unstranded'~'The protocol is unstranded.',
       TRUE~'Could not guess the protocol type!') -> conclusion
```

{{< callout "automated" >}}
`r conclusion` Check the table though!
{{< /callout >}}

<!-- include a references section -->

```{r show_list_references, opts.label='batch_chunk', child=system.file('chunks/list_references.Rmd', package='project.notes.tools')}
```

<!-- include a session information section -->

```{r record_saved_objects, opts.label='batch_chunk'}
#! names of object to be included in the table of hashes
saved_objects <- c()
```

```{r show_session_information, opts.label='batch_chunk', child=system.file('chunks/session_information.Rmd', package='project.notes.tools')}
```

<!-- end of session information -->

