---
title: {{ default (.File.Path | path.Dir | path.Base | replaceRE "^20\\d+-\\d+-\\d+-" "" | humanize | title) (getenv "TITLE") }}
weight: 2

categories:

tags:

authors:
  - christopher-barrington

type: docs
toc: true
draft: false

date: `r format(Sys.time(), "%Y-%m-%d")`
created: {{ now.Format "2006-01-02" }}

always_allow_html: true
output:
  html_document:
    keep_md: yes
    df_print: paged
---

<!-- initial set up chunks are invisible -->

<!-- read the standard knitr setup script -->
```{r knitr_setup, echo=FALSE, code=readLines(system.file('chunks/knitr.r', package='project.notes.tools'))}
```

<!-- make subdirectories to write into/any bash work -->
```{bash setup_wd, opts.label='batch_chunk', code=readLines(system.file('chunks/working_directory.sh', package='project.notes.tools'))}
```

<!-- load any packages for the workbook -->
```{r load_workbook_packages, opts.label='batch_chunk'}
library(DOSE)
library(clusterProfiler)
library(fgsea)
library(biomaRt)
library(DESeq2)
```

<!-- read the standard setup script, loading well-used libraries -->
```{r configure_r, opts.label='batch_chunk', code=readLines(system.file('chunks/environment.r', package='project.notes.tools'))}
```

<!-- DOIs for software used in the workbook or to generate presented results, a named vector of strings -->
```{r add_doi_to_bibliography, cache=TRUE, echo=FALSE}
c(clusterprofiler_4='10.1016/j.xinn.2021.100141',
  clusterprofiler='10.1089/omi.2011.0118',
  dose='10.1093/bioinformatics/btu684',
  fgsea='10.1101/060012',
  msigdb='10.1073/pnas.0506580102',
  msigdb_hallmark='10.1016/j.cels.2015.12.004') -> bibliography_doi
```

<!-- write a bibliography -->
```{r write_bibliography, dependson='add_doi_to_bibliography', cache=TRUE, echo=FALSE, code=readLines(system.file('chunks/write_bibliography.r', package='project.notes.tools'))}
```

<!-- set a ggplot theme to use as a base in the workbook -->
```{r define_global_ggplot_theme, opts.label='batch_chunk'}
(theme_bw() +
 theme(text=element_text(size=10, colour='black'),
       panel.grid.minor=element_blank(),
       strip.background=element_blank())) %>%
theme_set()
```

<!-- any final things to set for this workbook -->
```{r final_setup, opts.label='batch_chunk'}
```

<!-- define slugs for other reports here for ease and reuse -->
```{r define_slugs, opts.label='batch_chunk'}
slugs <- list(deseq2='deseq2')
```

<!-- define shortcut links -->
[CRICK]: https://www.crick.ac.uk

[clusterprofiler bioc]: https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html
[clusterprofiler do]: http://yulab-smu.top/clusterProfiler-book/chapter4.html#enrichdo-function
[clusterprofiler go]: http://yulab-smu.top/clusterProfiler-book/chapter5.html
[clusterprofiler kegg]: http://yulab-smu.top/clusterProfiler-book/chapter6.html
[fgsea]: http://bioconductor.org/packages/release/bioc/html/fgsea.html
[msigdb collections]: http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp

```{r write_slugs, opts.label='batch_chunk', code=readLines(system.file('chunks/write_slugs.r', package='project.notes.tools')), results='asis'}
```

<!-- end of initial set up chunks -->

Results from the [DESeq2 analysis][deseq2 post] will be used here to find gene ontologies and pathways that are enriched in differentially expressed genes. Direction-specific changes in comparisons will be tested using [`clusterProfiler`][clusterprofiler bioc] to examine databases such as Gene Ontology terms and KEGG pathways.

<!--more-->

## Load differentially expressed genes

The results of the [DESeq2 analysis][deseq2 post] are loaded here from the `all_dds` list of `dds` objects. The results in the `dds` are differentially expressed genes. The universe of possible genes is defined here as the set of genes in the `dds` that were detected in at least one dataset (and therefore have `allZero==FALSE`).

{{< callout "note" >}}
The list of results objects can be filtered to retain only comparisons for which differentially expressed genes were detected and we are interested in.
{{< /callout >}}

{{< callout "warning" >}}
The list of results is filtered here to remove LRT-based results. These results need to be assessed within grouped responses, rather the differentially expressed gene sets.
{{< /callout >}}

```{r load_deseq2_results}
data(all_dds)

#! filter for Wald-based comparisons
all_dds %>%
  keep(function(x) metadata(x) %>% names() %>% is_in(x='contrast')) %>%
  names() %>%
  magrittr::extract(all_dds, .) -> all_dds

#! filter by comparison name for relevant comparisons
all_dds %>%
  names() %>%
  str_detect(pattern='.*') %>%
  magrittr::extract(all_dds, .) -> all_dds

#! collect ids of differentially expressed genes
ldply(all_dds, .id='comparison', .parallel=TRUE, function(dds)
  dds %>% 
    metadata() %>%
    pluck('results') %>%
    filter(is_in(outcome, c('up*', 'down*'))) %>%
    mutate(outcome=str_remove(outcome, pattern='\\*$')) %>%
    left_join(y={dds %>% metadata() %>% pluck('gene_information') %>% select(ensembl_gene_id, gene_name)}) %>%
    select(comparison, ensembl_gene_id, gene_name, outcome, shrunk.log2FoldChange, stat)) -> significant_results
```

The number of differentially expressed genes considered for each comparison is shown below:

```{r show_number_of_de_genes_considered, dependson='load_deseq2_results', echo=FALSE}
#! summary table of number of de genes
significant_results %>%
  group_by(comparison, outcome) %>%
  tally() %>%
  spread(key=outcome, value=n) %>%
  mutate(total=up+down) %>%
  kable(format=if_else(interactive(), 'markdown', 'html'),
        format.args=list(big.mark=','))
```

The universe of genes is defined below. (In most cases all universes will be the same in a project...)

```{r define_universe, dependson='load_deseq2_results'}
llply(all_dds, .parallel=TRUE, function(dds)
  mcols(dds) %>%
    as.data.frame() %>%
    filter(allZero==FALSE) %>%
    rownames()) %>%
  Reduce(f=unique) -> universe
```

In total `r length(universe) %>% comma()` genes were identified in one or more dataset in the project.

The `significant_results` are reformatted here to make a list of comparison and direction of change (and an additional 'changed' direction).

```{r get_gene_lists, dependson='load_deseq2_results'}
significant_results %>%
  rbind(mutate(significant_results, outcome='changed')) %>%
  dlply(~comparison+outcome, pluck, 'ensembl_gene_id') -> de_genes
```

<!--  -->

## `clusterProfiler`

The [`clusterProfiler`][clusterprofiler bioc] package will be used to test for enrichment of gene ontology terms, KEGG pathways and disease ontology (DO) terms. The following sections create functions to run the tests.

<!--  -->

### Create an identifier conversion vector

For the KEGG pathway enrichment with `enrichKEGG` the Ensembl gene identifier must be converted, in this case, to a UniProt identifier. The following chunk uses the `biomaRt` object created for the project and creates a named vector of Ensembl and UniProt identifiers so an Ensembl ID can be converted to UniProt.

```{r get_ensembl_id_conversions}
data(mart)

getBM(attributes=c('ensembl_gene_id', 'uniprotswissprot'), mart=mart) %>%
  filter(uniprotswissprot!='') %>%
  deframe() -> ensembl_2_uniprot

getBM(attributes=c('ensembl_gene_id','entrezgene'), mart=mart) %>%
  filter(entrezgene!='') %>%
  deframe() -> ensembl_2_entrez
```

The `biomaRt` object connects to ``r mart@host %>% str_remove('/biomart/martservice')``.

<!--  -->

### Define enrichment test functions

A function to test for enrichment of [Gene Ontology terms][clusterprofiler go], [KEGG pathways][clusterprofiler kegg] or [Disease Ontology][clusterprofiler do] are defined here. These functions accept two lists of Ensembl IDs: one gene set and one background (or `universe`). The results returned are with a 1% significance threshold.

* `enrichGO` tests for enrichment of Gene Ontologies in differentially expressed genes
* `enrichDO` tests whether differentially expressed genes are linked to diseases (but requires conversion to human gene identifiers)
* `enrichKEGG` tests for enrichment of differentially expressed genes in KEGG pathways

```{r define_enrichment_test_functions}
test_GO_enrichment <- function(gene_ids, universe, org.db, ont='BP', cutoff=0.01, ...)
  enrichGO(gene=gene_ids,
           universe=universe,
           OrgDb=org.db,
           keyType='ENSEMBL', # expects Ensembl IDs
           ont=ont,
           pAdjustMethod='fdr',
           pvalueCutoff=cutoff,
           qvalueCutoff=cutoff)

test_KEGG_pathway_enrichment <- function(gene_ids, universe, organism, cutoff=0.01, ...)
  enrichKEGG(gene=ensembl_2_uniprot[gene_ids],
             organism=organism,
             keyType='uniprot',
             universe=ensembl_2_uniprot[universe],
             pAdjustMethod='fdr',
             pvalueCutoff=cutoff,
             qvalueCutoff=cutoff)

test_DO_enrichment <- function(gene_ids, universe, ont='DO', cutoff=0.01, ...)
  enrichDO(gene=ensembl_2_entrez[gene_ids],
           ont=ont,
           universe=ensembl_2_entrez[universe],
           pAdjustMethod='fdr',
           pvalueCutoff=cutoff,
           qvalueCutoff=cutoff)
```

<!--  -->

### Prepare for Slurm distribution

The enrichment tests are run in parallel using using `rslurm` to distribute each comparison across CAMP. This is required due to an error when running in parallel local threads where multiple threads accessing the `OrgDb` cause an exception (`database disk image is malformed`).

#### A function to run a test

This function is run on a node in a new R session. It will pick the list of genes and run one of the enrichment tests, returning an `enrichResult` object in a list named as the ID of the function's parameter set.

```{r function_to_run_on_slurm, dependson='define_enrichment_test_functions'}
run_on_slurm <- function(gene_set, enrichment_function, cutoff, ontology_type, id) {
  message('--------------------------------------------------')
  system('hostname')

  data(genome)

  list(gene_ids=de_genes[[gene_set]],
       universe=universe,
       ont=ontology_type,
       cutoff=cutoff,
       organism={genome$organism %>% str_trunc(width=3, side='right', ellipsis='')},
       org.db=genome$org.db) %>%
    do.call(what=get(enrichment_function)) %>%
    list() %>%
    set_names(id) -> result

  message('-- FINISHED')
  result
}
```

<!--  -->

#### Define enrichment test parameters

Each combination of test parameters is produced by `crossing()` the different parameter values that will be passed to the previously defined test functions. The 'gene_set' variable dictates which genes are tested. The parameter combinations (rows) are shuffled so that the computational cost of analyses are randomly distributed among nodes.

```{r define_enrichment_test_parameters, dependson='get_gene_lists'}
crossing(gene_set=names(de_genes),
         enrichment_function=c('test_GO_enrichment', 'test_KEGG_pathway_enrichment', 'test_DO_enrichment'),
         cutoff=0.01) %>%
  mutate(ontology_type=case_when(enrichment_function=='test_GO_enrichment'~'BP',
                                 enrichment_function=='test_DO_enrichment'~'DOLite',
                                 TRUE~'none'),
         id=seq(n())) %>%
  sample_frac(size=1) -> enrichment_test_parameters
```

<!--  -->

### Run enrichment tests

Using the above functions and parameters, the computation is distributed across CAMP. The Slurm job array is submitted (see note) and the results loaded once the job has finished.

```{r run_enrichment_tests, dependson=c('get_ensembl_id_conversions', 'function_to_run_on_slurm', 'define_enrichment_test_parameters'), results='hide'}
slurm_apply(f=run_on_slurm,
            params=enrichment_test_parameters,
            jobname='run_enrichment_tests',
            libPaths=.libPaths(),
            pkgs=c('DOSE', 'clusterProfiler', 'tidyverse', 'datarepository'),
            global_objects=c('de_genes', 'test_GO_enrichment', 'test_KEGG_pathway_enrichment', 'test_DO_enrichment', 'universe', 'ensembl_2_uniprot', 'ensembl_2_entrez'),
            nodes=1000, cpus_per_node=1, submit=TRUE,
            slurm_options=list(partition='cpu', `cpus-per-task`=2, mem='16G', time='00:30:00')) %T>%
  print() %>%
  saveRDS(file='scratch/slr_job.run_enrichment_tests.rds')
```

The Slurm job was:

```{r show_slurm_job, cache=FALSE, echo=FALSE}
if(file.exists('scratch/slr_job.run_enrichment_tests.rds')) {
  readRDS('scratch/slr_job.run_enrichment_tests.rds') %>%
    {.[c('jobname', 'jobid', 'nodes')]} %>%
    as.data.frame() %>%
    kable(format=if_else(interactive(), 'markdown', 'html'), align='lrr')
} else {
  knit_exit('no slurm job found!')
}
```

<!--  -->

### Load enrichment test results

The results of the tests are loaded into the `enrichment_results` `data.frame` and saved into the data repository. `xlsx` workbooks for each test with worksheets for each gene set [can be downloaded](files/enrichment_results.zip). These results are significant with the 1% `pvalue` (and `qvalue`) cutoff. I have applied the `clusterProfiler::simplify` function to try and make the results simpler to interpret.

```{r load_enrichment_test_results, dependson='run_enrichment_tests'}
readRDS('scratch/slr_job.run_enrichment_tests.rds') %>%
  get_slurm_out(outtype='raw', ncores=ncores, wait=TRUE) %>%
  unname() %>%
  unlist() %>%
  llply(.parallel=TRUE, function(obj)
    tryCatch(expr=clusterProfiler::simplify(obj), error=function(...) obj)) -> enrichResult_objects

enrichResult_objects %>%
  ldply(.id='id', as.data.frame) %>%
  mutate(id={as.character(id) %>% as.numeric()}) %>%
  left_join(y=enrichment_test_parameters, by='id') %>%
  select(-id) -> enrichment_results
```

```{r save_enrichment_results, dependson='load_enrichment_test_results', echo=FALSE, results='hide'}
save(enrichResult_objects)
save(enrichment_results)

enrichment_results %>%
  dlply(~enrichment_function, function(x)
    dlply(x, ~gene_set, select, -pvalue, -qvalue, -gene_set, -enrichment_function, -cutoff)) %>%
  Map(data=., filename=names(.), function(data, filename) {
    file.path('scratch/enrichment_results/') %T>%
      dir.create(recursive=TRUE, showWarnings=FALSE) %>%
      file.path(str_c(filename, '.xlsx')) %>%
      write.xlsx(x=data)}) %>%
  invisible()

system(command='cd scratch && zip - enrichment_results/*xlsx > ../files/enrichment_results.zip')
```

A summary of the number of enriched terms in each test, comparison and gene set is shown in the following table:

{{< tabs "show_enrichment_test_results_summary_tables" >}}
```{r show_enrichment_test_results_summary_tables, dependson='load_enrichment_test_results', echo=FALSE, results='asis'}
enrichment_results %>%
  group_by(enrichment_function, gene_set) %>%
  tally() %>%
  separate(col=gene_set, into=c('Comparison', 'Tx direction'), sep='\\.') %>%
  plyr::dlply(~enrichment_function, select, -enrichment_function) -> tables_list

for(i in names(tables_list)) {
  str_replace_all(i, '_', ' ') %>% str_to_title() %>% sprintf(fmt='{{< tab "%s" >}}') %>% cat()
  kable(x=tables_list[[i]], format=if_else(interactive(), 'markdown', 'html')) %>% show()
  cat('{{< /tab >}}')
}
```
{{< /tabs >}}

In the following excerpt table, the top 5 enriched terms ranked by adjusted p-value are shown for each contrast/direction combination. RNA-seq changes that were not direction-filtered (the 'changed' set) are not shown in the table but are included in [the Excel workbooks](files/enrichment_results.zip).

```{r enrichment_test_results_excerpt_table, dependson='load_enrichment_test_results', echo=FALSE}
enrichment_results %>%
  arrange(p.adjust) %>%
  filter(str_detect(gene_set, '\\.changed$')) %>%
  plyr::ddply(~enrichment_function+gene_set, head, n=5) %>%
  select(gene_set, enrichment_function, 1:4) %>%
  kable(format=if_else(interactive(), 'markdown', 'html'))
```

<!--  -->

## Gene set enrichment analysis

This analysis is similar to the ontology enrichment, this analysis looks for ranked gene changes in the complete results set to identify the magnitude of transcription change that is within different functional pathways. `fgsea` is then used to position the pathways on top of the `DESeq2` results which are ordered by their `stat` value. This analysis considers all genes for which there is expression data, not just the significantly different genes.

In the plots below, some pathways were selected. They show the gene set (any gene in the genome) ordered by `stat` from very positive (`rank=0`) to very negative, so the most highly changing genes are at the extreme positions. The line plot indicates the position of the genes in the pathway in the ordered gene set. The plots shown below are for pathways with `FDR<1%` (see table below).

### MSigDB pathways

[`fgsea`][fgsea] is used to test for enrichment of the [MSigDB hallmark pathways][msigdb collections], which is a curated set of genes for a well-studied set of pathways. The results in `xlsx` format are [available](files/fgsea_hallmark.xlsx).

First, the gene sets for database version 7.4 are sought from disk, and the genes in the pathway set loaded using `gmtPathway()` ready to be be used in `fgsea()` and later plotting. Pathway sets from the available collection are selected here ([others can be downloaded][msigdb collections]).

```{r show_available_msigdb_gene_sets, echo=FALSE}
file.path('/camp/stp/babs/working/barrinc', 'db', 'msigdb') %>%
  list.files(path=., pattern='v7.4.symbols.gmt') %>%
  data.frame(`gene set`=.) %>%
  kable(format=if_else(interactive(), 'markdown', 'html'))
```

```{r get_msigdb_gene_sets}
pathway_sets <- c('h.all')

file.path('/camp/stp/babs/working/barrinc', 'db', 'msigdb') %>%
  data.frame(path=., file=list.files(path=., pattern='v7.4.symbols.gmt')) %>%
  mutate(name=str_remove_all(file, '.v7.4|.symbols|.gmt')) %>%
  unite(col=file, sep='/', path, file) %>%
  rev() %>%
  deframe() %>%
  magrittr::extract(pathway_sets) %>%
  llply(gmtPathways) -> msigdb_pathway_sets
```

The genes in each comparison are now ranked by their `stat` value; this is calculated by DESeq2 and is the Wald statistic. (The [comparisons were filtered above](#load-differentially-expressed-genes) to remove LRT-based comparisons). The genes at the extremes of these vectors are likely to be differentially expressed, but there is no filtering here for significance.

```{r get_ranked_results, dependson='load_deseq2_results'}
ldply(all_dds, .id='comparison', .parallel=TRUE, function(dds)
  dds %>% 
    metadata() %>%
    pluck('results') %>%
    select(ensembl_gene_id, stat) %>%
    left_join(y={dds %>% metadata() %>% pluck('gene_information') %>% select(ensembl_gene_id, gene_name)}) %>%
    select(ensembl_gene_id, gene_name, stat)) %>%
  drop_na() %>%
  arrange(comparison, stat) %>%
  dlply(~comparison, select, gene_name, stat) %>%
  llply(deframe) -> ranked_results
```

The `ranked_results` is a list of comparison results, each of which is a named vector of genes and their `stat`. An excerpt is shown below.

```{r show_ranked_significant_results, dependson='get_ranked_results', echo=FALSE}
ranked_results %>%
  head(n=3) %>%
  lapply(headtail, n=5)
```

GSEA is now run using `fgsea` for every pathway set and comparison combination in a nested list in parallel. A [zip file of Excel workbooks](files/fgsea_results.xlsx.zip) is created that contains a workbook for each pathway set, each of which contains a worksheet for each comparison. 

```{r run_fgsea, dependson=c('get_msigdb_gene_sets','get_ranked_results'), out.width='50%', results='asis'}
llply(msigdb_pathway_sets, .parallel=TRUE, function(msigdb_pathway_set)
  llply(ranked_results, .parallel=TRUE, function(ranked_result)
    fgsea(pathways=msigdb_pathway_set, stats=ranked_result, nperm=5000, nproc=4))) %>%
  map_depth(.depth=2, arrange, padj, desc(size), pathway) -> fgsea_results
```

```{r write_fgsea_results_excel_workbooks, dependson='run_fgsea', echo=FALSE}
dir.create(path='scratch/fgsea_results', recursive=TRUE, showWarnings=FALSE)
system(command='rm -r scratch/fgsea_results/*')

Map(fgsea_result=fgsea_results, pathway_set=names(fgsea_results), f=function(fgsea_result, pathway_set)
  path('scratch/fgsea_results', pathway_set, ext='xlsx') %>%
    write.xlsx(x=fgsea_result)) %>%
  invisible()

system(command='cd scratch && zip - fgsea_results/*xlsx > ../files/fgsea_results.xlsx.zip')
```

### Plot `fgsea` results

The results from the `fgsea` analysis are plotted below.

First, a summary plot of the top pathways for up- or down-regulated (not necessarily significantly so) genes. The top 6 pathways at either extreme of the results ranked by ES are shown below.

{{< callout "wip" >}}
This needs some/extensive prettification.
{{< /callout >}}

```{r show_fgsea_pathways_table_plots, dependson='run_fgsea', fig.width=8.5, fig.height=6, echo=FALSE, results='asis'}
cat('{{< tabs "show_fgsea_pathways_table_plots" >}}')

for(msigdb_pathway_set in names(fgsea_results)) {
  sprintf(fmt='{{< tab "%s" >}}', msigdb_pathway_set) %>% cat()
  sprintf(fmt='{{< tabs "show_fgsea_pathways_table_plots-%s" >}}', msigdb_pathway_set) %>% cat()

  for(comparison in names(fgsea_results[[msigdb_pathway_set]])) {
    sprintf(fmt='{{< tab "%s" >}}', comparison) %>% cat()

    fgsea_results %>%
      pluck(msigdb_pathway_set, comparison) -> fgsea_result

    fgsea_result %>%
      arrange(desc(ES)) %>%
      helpr::headtail(n=6) %>%
      arrange(padj) %>%
      pluck('pathway') %>%
      magrittr::extract(msigdb_pathway_sets[[msigdb_pathway_set]], .) %>%
      plotGseaTable(pathways=.,
                    stats=ranked_results[[comparison]],
                    fgseaRes=fgsea_result,
                    gseaParam=1,
                    colwidths=c(4,2,1,0,1), render=FALSE) %>%
      show_newpage_grid()

    cat('{{< /tab >}}')
  }

  cat('{{< /tabs >}}')
  cat('{{< /tab >}}')
}

cat('{{< /tabs >}}')
```

Enrichment plots for the top 6 most extreme ES results are shown below. A [zip file of enrichment plots](files/output_fgsea_enrichment_plots.zip) for every pathway set and comparison combination is made because there are too many plots to include here.

```{r get_fgsea_enrichment_plots, dependson='run_fgsea'}
fgsea_results %>%
  Map(pathway_set=names(.), fgsea_results=., function(fgsea_results, pathway_set) {
    fgsea_results %>%
      Map(comparison=names(.), fgsea_result=., function(fgsea_result, comparison) {
        fgsea_result %>%
          arrange(desc(ES)) %>%
          arrange(padj) %>%
          pluck('pathway') %>%
          magrittr::extract(msigdb_pathway_sets[[pathway_set]], .) %>%
          llply(.parallel=TRUE, function(pathway)
            plotEnrichment(pathway=pathway, stats=ranked_results[[comparison]], gseaParam=1) +
              scale_x_continuous(labels=comma) +
              labs(x='Ranked genes', y='Enrichment score'))})}) -> plots_list
```

```{r output_fgsea_enrichment_plots, dependson='get_fgsea_enrichment_plots', results='hide', echo=FALSE}
file.path('scratch', 'output_fgsea_enrichment_plots') %>%
  when(dir.exists(.)~sprintf(fmt='rm -r %s', .),
       TRUE~'sleep 0') %>%
  system()

Map(pathway_set=names(plots_list), plots_list=plots_list, function(plots_list, pathway_set)
  Map(comparison=names(plots_list), plots_list=plots_list, function(plots_list, comparison)
    Map(pathway=names(plots_list), enrichment_plot=plots_list, function(enrichment_plot, pathway) {
      path('scratch', 'output_fgsea_enrichment_plots', pathway_set, comparison, pathway, ext='pdf') %T>%
        {dir.create(path=dirname(.), recursive=TRUE, showWarnings=FALSE)} -> filename

      enrichment_plot %>%
        helpr:::set_panel_dims(width=unit(3, 'in'), height=unit(3/1.6, 'in')) %>%
        ggsave(filename=filename, width=15, height=15, units='in')

      sprintf(fmt='pdfcrop %s %s', filename, filename) %>% system()}))) %>%
  invisible()

system(command='cd scratch && zip -r - output_fgsea_enrichment_plots > ../files/output_fgsea_enrichment_plots.zip')
```

```{r show_fgsea_enrichment_plots, dependson='get_fgsea_enrichment_plots', echo=FALSE, results='asis'}
cat('{{< tabs "show_fgsea_enrichment_plots" >}}')

for(msigdb_pathway_set in {plots_list %>% names()}) {
  sprintf(fmt='{{< tab "%s" >}}', msigdb_pathway_set) %>% cat()
  sprintf(fmt='{{< tabs "show_fgsea_enrichment_plots-%s" >}}', msigdb_pathway_set) %>% cat()

  for(comparison in {plots_list %>% pluck(msigdb_pathway_set) %>% names()}) {
    sprintf(fmt='{{< tab "%s" >}}', comparison) %>% cat()
    sprintf(fmt='{{< tabs "show_fgsea_enrichment_plots-%s-%s" >}}', msigdb_pathway_set, comparison) %>% cat()

    for(pathway in {plots_list %>% pluck(msigdb_pathway_set, comparison) %>% names() %>% headtail(n=3)}) {
      sprintf(fmt='{{< tab "%s" >}}', pathway) %>% cat()

      plots_list %>%
        pluck(msigdb_pathway_set, comparison, pathway) %>%
        resize_and_show(width=3)
  
      cat('{{< /tab >}}')
    }

    cat('{{< /tabs >}}')
    cat('{{< /tab >}}')

  }

  cat('{{< /tabs >}}')
  cat('{{< /tab >}}')
}

cat('{{< /tabs >}}')
```

<!-- a methods section for a manuscript, use {{< textcite "author-year" >}} -->

## Methods section

> Enrichment of gene function was assessed in R (`r R.version[c('major','minor')] %>% unlist() %>% str_c(collapse='.')`) using clusterProfiler (`r packageVersion('clusterProfiler')`, {{< textcite "Wu-2021" >}}, {{< textcite "Yu-2012" >}}) to query the Gene Ontology, KEGG and DOSE ({{< textcite "Yu-2014" >}}) databases. The fgsea (`r packageVersion('fgsea')`) R package was used for gene set enrichment analysis using all detected genes ranked by the stat variable of the DESeq2 results and pathway sets defined by MSigDB ({{< textcite "Subramanian-2005" >}}, {{< textcite "Liberzon-2015" >}}). 

<!-- include a references section -->

```{r show_list_references, opts.label='batch_chunk', child=system.file('chunks/list_references.Rmd', package='project.notes.tools')}
```

<!-- include a session information section -->

```{r record_saved_objects, opts.label='batch_chunk'}
#! names of object to be included in the table of hashes
saved_objects <- c('enrichResult_objects', 'enrichment_results')
```

```{r show_session_information, opts.label='batch_chunk', child=system.file('chunks/session_information.Rmd', package='project.notes.tools')}
```

<!-- end of session information -->
